xds/internal/xdsclient/pool.go

```diff
@@ -0,0 +1,143 @@
+// pool.
+//
+// The second return value represents a close function which the caller is
+// expected to invoke once they are done using the client.  It is safe for the
+// caller to invoke this close function multiple times.
+func (p *Pool) NewClient(name string) (XDSClient, func(), error) {
+	config, err := bootstrap.GetConfiguration()
+	if err != nil {
+		return nil, nil, fmt.Errorf("xds: failed to get xDS bootstrap config: %v", err)
+	}
+	return newRefCounted(name, p, config, defaultWatchExpiryTimeout, defaultIdleChannelExpiryTimeout, backoff.DefaultExponential.Backoff)
+}
+
+// NewClientForTesting returns an xds client configured with the provided
+// options from the pool. If the client doesn't already exist, it creates a new
+// xds client and adds it to the pool.
+//
+// The second return value represents a close function which the caller is
+// expected to invoke once they are done using the client.  It is safe for the
+// caller to invoke this close function multiple times.
+//
+// # Testing Only
+//
+// This function should ONLY be used for testing purposes.
+func (p *Pool) NewClientForTesting(opts OptionsForTesting) (XDSClient, func(), error) {
+	if opts.Name == "" {
+		return nil, nil, fmt.Errorf("opts.Name field must be non-empty")
+	}
+	if opts.WatchExpiryTimeout == 0 {
+		opts.WatchExpiryTimeout = defaultWatchExpiryTimeout
+	}
+	if opts.IdleChannelExpiryTimeout == 0 {
+		opts.IdleChannelExpiryTimeout = defaultIdleChannelExpiryTimeout
+	}
+	if opts.StreamBackoffAfterFailure == nil {
+		opts.StreamBackoffAfterFailure = defaultStreamBackoffFunc
+	}
+	if opts.Contents != nil {
+		config, err := bootstrap.NewConfigForTesting(opts.Contents)
+		if err != nil {
+			return nil, nil, err
+		}
+		return newRefCounted(opts.Name, p, config, opts.WatchExpiryTimeout, opts.IdleChannelExpiryTimeout, opts.StreamBackoffAfterFailure)
+	}
+
+	return newRefCounted(opts.Name, p, p.config, opts.WatchExpiryTimeout, opts.IdleChannelExpiryTimeout, opts.StreamBackoffAfterFailure)
+}
+
+// GetClientForTesting returns an xds client created earlier using the given
+// name from the pool.
+//
+// The second return value represents a close function which the caller is
+// expected to invoke once they are done using the client.  It is safe for the
+// caller to invoke this close function multiple times.
+//
+// # Testing Only
+//
+// This function should ONLY be used for testing purposes.
+func (p *Pool) GetClientForTesting(name string) (XDSClient, func(), error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	c, ok := p.clients[name]
+	if !ok {
+		return nil, nil, fmt.Errorf("xDS client with name %q not found", name)
+	}
+	c.incrRef()
+	return c, grpcsync.OnceFunc(func() { clientRefCountedClose(name, p) }), nil
+}
+
+// SetFallbackBootstrapConfig to specify a bootstrap configuration to use a
+// fallback when the bootstrap env vars are not specified.
+func (p *Pool) SetFallbackBootstrapConfig(bootstrapContents []byte) error {
```

Comment:
This should also be changed to accept a config instead of JSON bootstrap contents.

Once we move to the new generic xDS client, all client creation will accept a config struct and not JSON bootstrap.

Reply:
Done